torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -280,17 +311,17 @@ void DistEngine::computeDependencies(
 }
 
 void DistEngine::execute_graph_task_until_ready_queue_empty(
-    const std::shared_ptr<GraphTask>& graph_task,
-    std::shared_ptr<Node> root_to_execute,
+    NodeTask&& node_task,
     bool incrementOutstandingTasks) {
   engine_.initialize_device_threads_pool();
   // Create a ready queue per call to traverse the graph_task from
   // root_to_execute This allow concurrent execution of the same GraphTask from
   // different threads
   std::shared_ptr<ReadyQueue> cpu_ready_queue = std::make_shared<ReadyQueue>();
-  cpu_ready_queue->push(
-      NodeTask(graph_task, std::move(root_to_execute), InputBuffer(0)),
-      incrementOutstandingTasks);
+  auto graph_task = node_task.base_.lock();
+  TORCH_INTERNAL_ASSERT(graph_task);
```

Comment:
Is this always true?
For the new use case, you keep the graph_task alive in the parent scope so it's fine.
But is it guaranteed in other cases as well? Or is it fine if another thread encounters an error and destroys the graph_task before you get here?